libgadget/density.c

```diff
@@ -202,29 +199,63 @@ density(int update_hsml, int DoEgyDensity, ForceTree * tree)
 
     walltime_measure("/SPH/Density/Init");
 
-    DENSITY_GET_PRIV(tw)->NPLeft = ta_malloc("NPLeft", int, All.NumThreads);
+    int NumThreads = omp_get_max_threads();
+    DENSITY_GET_PRIV(tw)->NPLeft = ta_malloc("NPLeft", size_t, NumThreads);
+    DENSITY_GET_PRIV(tw)->NPRedo = ta_malloc("NPRedo", int *, NumThreads);
+    int alloc_high = 0;
+    int * ReDoQueue = ActiveParticle;
+    int size = SlotsManager->info[0].size + SlotsManager->info[5].size;
+    if(size > NumActiveParticle)
+        size = NumActiveParticle;
 
     /* we will repeat the whole thing for those particles where we didn't find enough neighbours */
     do {
-        memset(DENSITY_GET_PRIV(tw)->NPLeft, 0, sizeof(int)*All.NumThreads);
+        int * CurQueue = ReDoQueue;
+
+        int tsize = size / NumThreads + 2;
+        /* The ReDoQueue swaps between high and low allocations so we can have two allocated alternately*/
+        if(update_hsml) {
+            if(!alloc_high) {
+                ReDoQueue = (int *) mymalloc2("ReDoQueue", tsize * sizeof(int) * NumThreads);
+                alloc_high = 1;
+            }
+            else {
+                ReDoQueue = (int *) mymalloc("ReDoQueue", tsize * sizeof(int) * NumThreads);
+                alloc_high = 0;
+            }
+            gadget_setup_thread_arrays(ReDoQueue, DENSITY_GET_PRIV(tw)->NPRedo, DENSITY_GET_PRIV(tw)->NPLeft, tsize, NumThreads);
```

Comment:
BTW, what was our verdict on jemalloc?

Reply:
Sorry, I didn't get a chance to look into it.

Reply:
OK. It looks like jemalloc doesn't record the line number of memory allocation / tag; so it is mostly out -- our allocation are mostly 'huge allocation's, an it seems jemalloc is mostly optimized for small alloc, and may actually perform bad regarding segmentation. 

But I think I found a way to avoid requiring exact ordering of malloc and free; and perhaps also a way to record full stack information when an allocation occurs. I'll think about this more.

Reply:
Ok not jemalloc then. One related thing that occurred to me is to generalise the algorithm in sfr_eff.c to allow realloc() to succeed even when called on not the top allocation.

Reply:
As long as we ensure the heap state after a time-step is identical then there is no long term fragmentation. So we don't really need ordered alloc/free. Time for a new allocator :-)

Reply:
hooray! :-)